flutter/lib/src/sentry_asset_bundle.dart

```diff
@@ -0,0 +1,204 @@
+///
+/// The `enableStructureDataTracing` setting is an experimental feature.
+/// Use at your own risk.
+class SentryAssetBundle implements AssetBundle {
+  SentryAssetBundle({
+    Hub? hub,
+    AssetBundle? bundle,
+    bool enableStructuredDataTracing = false,
+  })  : _hub = hub ?? HubAdapter(),
+        _bundle = bundle ?? rootBundle,
+        _enableStructuredDataTracing = enableStructuredDataTracing;
+
+  final Hub _hub;
+  final AssetBundle _bundle;
+  final bool _enableStructuredDataTracing;
+
+  @override
+  Future<ByteData> load(String key) async {
+    final span = _hub.getSpan()?.startChild(
+          'file.read',
+          description: 'AssetBundle.load: ${_fileName(key)}',
+        );
+
+    span?.setData('file.path', key);
+
+    ByteData? data;
+    try {
+      data = await _bundle.load(key);
+      _setDataLength(data, span);
+      span?.status = SpanStatus.ok();
+    } catch (exception) {
+      span?.throwable = exception;
+      span?.status = SpanStatus.internalError();
+      rethrow;
+    } finally {
+      await span?.finish();
+    }
+    return data;
+  }
+
+  @override
+  Future<T> loadStructuredData<T>(String key, _Parser<T> parser) {
+    if (_enableStructuredDataTracing) {
+      return _loadStructuredDataWithTracing(key, parser);
+    }
+    return _bundle.loadStructuredData(key, parser);
+  }
+
+  Future<T> _loadStructuredDataWithTracing<T>(
+      String key, _Parser<T> parser) async {
+    final span = _hub.getSpan()?.startChild(
+          'file.read',
+          description: 'AssetBundle.loadStructuredData<$T>: ${_fileName(key)}',
+        );
+    span?.setData('file.path', key);
+
+    final completer = Completer<T>();
+
+    // This future is intentionally not awaited. Otherwise we deadlock with
+    // the completer.
+    // ignore: unawaited_futures
+    runZonedGuarded(() async {
+      final data = await _bundle.loadStructuredData(
+        key,
+        (value) async => await _wrapParsing(parser, value, key, span),
+      );
+      span?.status = SpanStatus.ok();
+      completer.complete(data);
+    }, (exception, stackTrace) {
+      completer.completeError(exception, stackTrace);
+    });
+
+    T data;
+    try {
+      data = await completer.future;
+      _setDataLength(data, span);
+      span?.status = const SpanStatus.ok();
+    } catch (e) {
+      span?.throwable = e;
+      span?.status = const SpanStatus.internalError();
+      rethrow;
+    } finally {
+      await span?.finish();
+    }
+    return data;
+  }
+
+  @override
+  Future<String> loadString(String key, {bool cache = true}) async {
+    final span = _hub.getSpan()?.startChild(
+          'file.read',
+          description: 'AssetBundle.loadString: ${_fileName(key)}',
+        );
+
+    span?.setData('file.path', key);
+    span?.setData('from-cache', cache);
+
+    String? data;
+    try {
+      data = await _bundle.loadString(key, cache: cache);
+      span?.status = SpanStatus.ok();
```

Comment:
l: I believe we should set data 'file.size' here as well.

Reply:
I thought about that, too, but I don't believe the cost of doing it is worth it. We would have to convert the string back to bytes which is potentially an expensive operation.

Reply:
agreed, and if we do, we should consider what Flutter does too:

https://github.com/flutter/flutter/blob/5112f86c9416143a33de4ead42c8aff6153767f7/packages/flutter/lib/src/services/asset_bundle.dart#L75-L82

Reply:
It make sense